src/fiffs.cpp

```diff
@@ -136,46 +129,47 @@ static int reply_buf_limited(fuse_req_t req, const char *buf, size_t bufsize, of
     debug_printf("fiffs_readdir\n");
 
-    if (ino != 1)
-        fuse_reply_err(req, ENOTDIR);
-    else {
-        size_t bufsize;
-
-        std::vector<int> entry_sizes(FS::inodes.size());
-        entry_sizes[0] = fuse_add_direntry(req, NULL, 0, ".", NULL, 0);
-        entry_sizes[1] = fuse_add_direntry(req, NULL, 0, "..", NULL, 0);
-
-        bufsize = entry_sizes[0] + entry_sizes[1];
-        for (int i = 2; i < FS::inodes.size(); ++i) {
-            entry_sizes[i] = fuse_add_direntry(req, NULL, 0, FS::inodes[i].name.c_str(), NULL, 0);
-            bufsize += entry_sizes[i];
-        }
-
-        struct stat stbuf;
-        memset(&stbuf, 0, sizeof(stbuf));
-        stbuf.st_ino = 1;
-        std::vector<char> buf(bufsize);
-
-        int offset = 0;
-        fuse_add_direntry(req, buf.data(), entry_sizes[0], ".", &stbuf, offset + entry_sizes[0]);
-        offset += entry_sizes[0];
-        fuse_add_direntry(req, buf.data() + offset, entry_sizes[1], "..", &stbuf, offset + entry_sizes[1]);
-        offset += entry_sizes[1];
-        for (int i = 2; i < FS::inodes.size(); ++i) {
-            stbuf.st_ino = i;
-            const auto &e_size = entry_sizes[i];
-            fuse_add_direntry(req, buf.data() + offset, e_size, FS::inodes[i].name.c_str(), &stbuf, offset + e_size);
-            offset += e_size;
-        }
-
-        reply_buf_limited(req, buf.data(), bufsize, off, size);
+    size_t bufsize;
+
+    std::list<int> children = FS::inodes[ino].children;
+    std::vector<int> entry_sizes(children.size() + 2);
+    entry_sizes[0] = fuse_add_direntry(req, NULL, 0, ".", NULL, 0);
+    entry_sizes[1] = fuse_add_direntry(req, NULL, 0, "..", NULL, 0);
+
+    bufsize = entry_sizes[0] + entry_sizes[1];
+    int i = 2;
+    for (const auto &cno : children) {
+        entry_sizes[i] = fuse_add_direntry(req, NULL, 0, FS::inodes[cno].name.c_str(), NULL, 0);
+        bufsize += entry_sizes[i];
+        i++;
+    }
+
+    struct stat stbuf;
+    memset(&stbuf, 0, sizeof(stbuf));
+    stbuf.st_ino = ino;
+    std::vector<char> buf(bufsize);
+
+    int offset = 0;
+    fuse_add_direntry(req, buf.data(), entry_sizes[0], ".", &stbuf, offset + entry_sizes[0]);
+    offset += entry_sizes[0];
+    fuse_add_direntry(req, buf.data() + offset, entry_sizes[1], "..", &stbuf, offset + entry_sizes[1]);
+    offset += entry_sizes[1];
+    i = 2;
+    for (const auto &cno : children) {
+        stbuf.st_ino = FS::inodes[cno].st.st_ino;
+        const auto &e_size = entry_sizes[i];
+        fuse_add_direntry(req, buf.data() + offset, e_size, FS::inodes[cno].name.c_str(), &stbuf, offset + e_size);
+        offset += e_size;
+        i++;
     }
+
+    reply_buf_limited(req, buf.data(), bufsize, off, size);
 }
 
 static void fiffs_open(fuse_req_t req, fuse_ino_t ino, struct fuse_file_info *fi) {
     debug_printf("fiffs_open\n");
-    if (ino < 2)
-        fuse_reply_err(req, EISDIR);
+    if (ino < 1 || ino > FS::inodes.size())
```

Comment:
This allows for opens on non-existent inodes (0 indexed). We should probably also add a check on if the inode is actually a file, since you can't open a directory.

Reply:
Thanks. Hm, I guess these don't exactly correspond to syscalls. I see fuse has an opendir like opendir(3). Normally open(2) will open a directory so that you can readdir(3).

Reply:
I just added opendir. I guess there's no readdir(2), so it makes sense